### DISCUSSION TITLE
How to share a data protection key among tenants?

### DISCUSSION THREAD

**Discussion by hieucd04** [OrchardCMS/OrchardCore#12356] [Q&A] [answered]
I have 2 Orchard Core tenants:
   - Tenant A: the default tenant under domain name `www.localhost.com`
   - Tenant B: a tenant which exposes an API under domain name `api.localhost.com`

At tenant A, I issues a cookie with `domain` field set to `.localhost.com`.
Now, at tenant B, I want to reuse the above cookie to authenticate my API (Single Sign-On).
The problem is: I can't decrypt that cookie because each tenant has its own data protection key.

And to make the problem worst, I looked into Orchard Core source code and found this:
https://github.com/OrchardCMS/OrchardCore/blob/4093d2e71fc9ef48c96afd609a6564826e17b803/src/OrchardCore/OrchardCore/Modules/Extensions/ServiceCollectionExtensions.cs#L434-L437

Does it mean there are no hope left for me?

**Comment by Skrypt** [comment, marked as answer]
Use OpenId module for single sign on.

**Reply by ns8482e** [reply]
True

 @hieucd04  the approach you are taking is opposite of idea of tenant isolation

if you need tenant isolation then you need to use open id as @Skrypt suggest

if you don’t need tenant isolation but need to use more than one domain that shares data then use single tenant and  provide  host names

**Reply by hieucd04** [reply]
I totally forgot that `OpenId` supports this.
Thank you! 😄

**Comment by jtkech** [comment]
Or use a shared data protection, for example our `OC.Redis.DataProtection` feature (needs a running Redis instance).

**Reply by ns8482e** [reply]
@jtkech having shared keyring between tenants will also not help here, as encryption is based on tenant name :)
And  you can’t have duplicate tenants 

may need deployment on two different nodes with same tenant name

**Reply by jtkech** [reply]
>  as encryption is based on tenant name :)

Are you sure? ;)

**Reply by ns8482e** [reply]
https://github.com/OrchardCMS/OrchardCore/blob/4093d2e71fc9ef48c96afd609a6564826e17b803/src/OrchardCore/OrchardCore/Modules/Extensions/ServiceCollectionExtensions.cs#L441

**Reply by jtkech** [reply]
Oops yes, didn't remember that it allows to share data protection across OC instances but for the same tenant ;)

**Reply by hieucd04** [reply]
> (needs a running Redis instance)

While this post directly answers my question it is unfortunately not viable in my case because of this.
